src/services/outliningElementsCollector.ts

```diff
@@ -89,6 +107,60 @@ namespace ts.OutliningElementsCollector {
             return isFunctionBlock(node) && node.parent.kind !== SyntaxKind.ArrowFunction;
         }
 
+        function getRegionName(start: number, end: number) {
+            if (!isInComment(sourceFile, start)) {
+                const comment = sourceFile.text.substring(start, end).trim();
+                const result = comment.match(regionStart);
+
+                if (result && result.length > 0) {
+                    const label = result.pop();
+                    if (label) {
+                        return label.trim();
+                    }
+                    else {
+                        return defaultLabel;
+                    }
+                }
+            }
+            return "";
+        }
+
+        function isRegionEnd(start: number, end: number) {
+            if (!isInComment(sourceFile, start)) {
+                const comment = sourceFile.text.substring(start, end).trim();
+                return !!comment.match(regionEnd);
+            }
+            return false;
+        }
+
+        function gatherRegions(): void {
+            const lineStarts = sourceFile.getLineStarts();
+
+            for (let i = 0; i < lineStarts.length; i++) {
+                const currentLineStart = lineStarts[i];
+                const lineEnd = lineStarts[i + 1] - 1 || sourceFile.getEnd();
+
+                const name = getRegionName(currentLineStart, lineEnd);

```

Comment:
looks like you are calling `isInComment` a few times every line. I would say this is more expensive than the other checks, which is basically does the line start with `//` or not. 

consider making this check first. then after that run your regexp, and if that all passes, then call `isInComment` before adding the region.

Reply:
Also consider combining your regexps into one. this way you run the regexp and gets you name + end/start marker.

Reply:
Combined the regexp and moved the `isInComment` after we know it's a relevant line.